sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -810,6 +784,35 @@ mod checked {
                 written_objects.insert(id, object);
             }
 
+            // Before finishing, ensure that any shared object taken by value by the transaction is either:
+            // 1. Mutated (and still has a shared ownership); or
+            // 2. Deleted.
+            // Otherwise, the shared object operation is not allowed and we fail the transaction.
+            for id in &by_value_shared_objects {
+                match written_objects.get(id) {
+                    Some(obj) => {
+                        if !obj.is_shared() {
+                            return Err(ExecutionError::new(
+                                ExecutionErrorKind::SharedObjectOperationNotAllowed,
+                                Some(
+                                    format!("Shared object operation on {} not allowed", id).into(),
```

Comment:
Can we make this errors more descriptive as to which operation is not allowed (e.g. in this case, the shared object can't be frozen, transferred or wrapped and in the case below it must be deleted since it wasn't re-shared)?

Reply:
Also, is this the right place to perform these checks? The reason I ask is that this won't cover system transactions, which are also allowed to take shared objects by value.  The alternative I can think of is to do this check when converting the `TemporaryStore` to effects.

Reply:
I believe every system transaction does run through this as well (unless I'm missing somewhere). Every system transaction that I can see calls into `programmable_transactions::execution::execute::<...>(...)` which always calls `finish` (this function).

Reply:
I think that's mostly correct, but for example change epoch safe mode won't use this code path, but does modify state on-chain.

Reply:
Out of curiosity where does the complication come from? (Naively I had hoped that `TemporaryStore` kept track of the shared object inputs so could tell, when converting to effects, what the issue was).  The safe mode example was more a "for instance" that we have ways for system transactions to create effects that don't flow through the PTB code.

Reply:
RE: system transactions -- as far as I can tell from looking through the code, the safe mode epoch advancement is the only system transaction that does skip the execution runtime (and it does so for a reason -- since we should continue to be able to advance the epoch no matter what pretty much).

As far as complications, outside of this place we don't actually know if the shared object has been used by value in the transaction or not. We could write more general rules on the "if a shared object occurs mutably in the inputs then.." type checks in the temporary store, but you would then quickly bump into this type of situation where it may then affect the safe mode epoch advancement that we don't want. The only way to get around that then would be to do a bunch of (more complex) logic, whereas where it is at now I believe handles each of these cases.

Also, happy to discuss synchronously more on Monday -- not planning on landing this until then anyways :)